### DISCUSSION TITLE
Error 503 InvalidConfigException for ldap provider

### DISCUSSION THREAD

**Discussion by mhdaouas** [directus/directus#10772] [Archive]
### Preflight Checklist

- [X] I have completed all [Troubleshooting Steps](https://docs.directus.io/getting-started/support/#troubleshooting-steps).
- [X] I'm on [the latest version of Directus](https://github.com/directus/directus/releases).
- [X] There's [no other issue](https://github.com/directus/directus/issues?q=is%3Aissue) that already describes my problem.

### Describe the Bug

I get an error when I try to authenticate with ldap.

I've added this in .env file:
```
AUTH_PROVIDERS="ldap"

AUTH_LDAP_DRIVER="ldap"
AUTH_LDAP_CLIENT_URL="ldap://..."
AUTH_LDAP_BIND_DN="CN=..."
AUTH_LDAP_BIND_PASSWORD="..."
AUTH_LDAP_USER_DN="..."
AUTH_LDAP_GROUP_DN="..."
```

Should I install a library like ldapjs manually or something? I don't get it.

### To Reproduce

See above

### What version of Directus are you using?

9.4.1

### What version of Node.js are you using?

v14.16.0

### What database are you using?

PostgreSQL 13

### What browser are you using?

Chrome

### What operating system are you using?

Windows 10

### How are you deploying Directus?

Running locally with VisualCode

**Comment by aidenfoxx** [comment]
Your bind user or password is probably invalid. Verify that the bind user details are correct.

**Comment by mhdaouas** [comment]
You mean BIND_DN or USER_DN?

**Comment by aidenfoxx** [comment]
`BIND_DN` and `BIND_PASSWORD` should be the authentication details of a user that can "query users and groups to perform authentication". The `BIND_DN` value should be the fully qualified DN of the user, including CN.

If the bind user is invalid, when you try and authenticate you will get an `InvalidConfigException`.

**Reply by mhdaouas** [reply]
how can I use LDAP filters in the env file?

**Reply by aidenfoxx** [reply]
There is currently no support for custom filters. You're welcome to create a PR to add them if the `_SCOPE` config options don't solve your issue.

**Reply by mhdaouas** [reply]
How to use the _SCOPE config options for filters? I checked in the doc but couldn't find anything helpful.

**Reply by aidenfoxx** [reply]
It affects the depth that LDAP searches. As specified in the docs:

```
[2] The scope defines the following behaviors:

    base: Limits the scope to a single object defined by the associated DN.
    one: Searches all objects within the associated DN.
    sub: Searches all objects and sub-objects within the associated DN.
```

**Reply by mhdaouas** [reply]
It doesn't work for me (even with sub).

**Reply by aidenfoxx** [reply]
You've not said what you're trying to achieve, or even if the `InvalidConfigException` has been fixed? It's very hard to help without knowing what you're trying to do.

**Reply by mhdaouas** [reply]
The `InvalidConfigException` has been fixed but now I try to authenticate users who need filtering without success (I get `request errored POST 503 /auth/login/ldap`).

I successfully could authenticate with ldapjs manually before (without using the Directus LDAP provider) with the following code:

```
var opts = {
	filter: "(&(objectClass=user))",
	filter: '(|(mail={' + <email adress>.split('@')[0] + '})(samaccountname={' + <email adress> + '})(mailNickName={' + <email adress> + '}))',
	scope: "sub",
	client: "*"
};

client.search('DC=<company>,DC=local', opts, function (err, res) { ...
```

**Reply by aidenfoxx** [reply]
You use `AUTH_<PROVIDER>_USER_ATTRIBUTE` to specify which attribute to use when authenticating. I assume you need it to be either `samaccountname`, `mail` or `mailNickName`. You can't specify multiple.

This is specified in the docs here, where all configuration options are defined: https://docs.directus.io/configuration/config-options/#ldap-ldap

If you need any additional functionality I recommend creating a PR to add it.

**Comment by mhdaouas** [comment]
I tried making changes (added filters) in ldap.ts under `api/src/auth/drivers` even though I know it's not a clean solution but couldn't make it work. I couldn't even make a `logger.warn()` work. Is that file used at runtime or should I make a modification somewhere else?

**Comment by mhdaouas** [comment]
Anyone can help?
